File: internal/application/clientcrypto/padding.go
package clientcrypto

import (
	"crypto/subtle"
	"errors"
	"fmt"
)

// PaddingBlockSize matches web/src/shared/crypto/padding.ts — item and folder
// names are padded to a multiple of 32 bytes before encryption to avoid
// leaking plaintext length.
const PaddingBlockSize = 32

// ErrInvalidPadding is returned when PKCS#7 verification fails.
var ErrInvalidPadding = errors.New("clientcrypto: invalid PKCS#7 padding")

// Pad appends PKCS#7 padding to data so the output length is a multiple of
// PaddingBlockSize. If data is already aligned, a full block (32 bytes) of
// padding is appended — standard PKCS#7 behaviour.
func Pad(data []byte) []byte {
	padLen := PaddingBlockSize - (len(data) % PaddingBlockSize)
	padded := make([]byte, len(data)+padLen)
	copy(padded, data)
	for i := len(data); i < len(padded); i++ {
		padded[i] = byte(padLen)
	}
	return padded
}

// Unpad strips PKCS#7 padding and verifies it in constant time. The input
// must be a non-empty multiple of PaddingBlockSize.
func Unpad(padded []byte) ([]byte, error) {
	if len(padded) == 0 || len(padded)%PaddingBlockSize != 0 {
		return nil, fmt.Errorf("%w: length %d not a multiple of %d", ErrInvalidPadding, len(padded), PaddingBlockSize)
	}
	padLen := int(padded[len(padded)-1])
	if padLen < 1 || padLen > PaddingBlockSize || padLen > len(padded) {
		return nil, fmt.Errorf("%w: byte %d out of range", ErrInvalidPadding, padLen)
	}
	// Constant-time verification of every padding byte.
	expected := make([]byte, padLen)
	for i := range expected {
		expected[i] = byte(padLen)
	}
	if subtle.ConstantTimeCompare(padded[len(padded)-padLen:], expected) != 1 {
		return nil, ErrInvalidPadding
	}
	return padded[:len(padded)-padLen], nil
}

File: web/src/shared/crypto/padding.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

/**
 * PKCS#7-style padding to next 32-byte boundary (M5).
 *
 * Applied to item names and folder names BEFORE encryption so that
 * ciphertext lengths don't leak plaintext length info.
 *
 * Each byte of padding has the value equal to the number of padding bytes.
 * If the input is already a multiple of 32, a full block of 32 padding bytes
 * is added (standard PKCS#7 behavior).
 */

const BLOCK_SIZE = 32;

/** Pad plaintext to the next 32-byte boundary using PKCS#7. */
export function pad(data: Uint8Array): Uint8Array {
  const padLen = BLOCK_SIZE - (data.length % BLOCK_SIZE);
  const padded = new Uint8Array(data.length + padLen);
  padded.set(data, 0);
  padded.fill(padLen, data.length);
  return padded;
}

/** Remove PKCS#7 padding. Throws on invalid padding. */
export function unpad(padded: Uint8Array): Uint8Array {
  if (padded.length === 0 || padded.length % BLOCK_SIZE !== 0) {
    throw new Error(
      `padding: invalid padded length ${padded.length} (must be non-zero multiple of ${BLOCK_SIZE})`,
    );
  }

  const padLen = padded[padded.length - 1]!;
  if (padLen < 1 || padLen > BLOCK_SIZE) {
    throw new Error(`padding: invalid pad byte 0x${padLen.toString(16)}`);
  }

  if (padLen > padded.length) {
    throw new Error("padding: pad length exceeds data length");
  }

  // Verify all padding bytes have the correct value
  for (let i = padded.length - padLen; i < padded.length; i++) {
    if (padded[i] !== padLen) {
      throw new Error("padding: inconsistent pad bytes");
    }
  }

  return padded.slice(0, padded.length - padLen);
}
